parser/src/token.rs

```diff
@@ -0,0 +1,669 @@
+use base::ast::is_operator_char;
+use base::pos::{self, BytePos, Column, Line, Location, Spanned};
+use std::str::Chars;
+use std::fmt;
+
+use self::ErrorCode::*;
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Token<'input> {
+    Identifier(&'input str),
+    Operator(&'input str),
+
+    StringLiteral(String),
+    CharLiteral(char),
+    IntLiteral(i64),
+    ByteLiteral(u8),
+    FloatLiteral(f64),
+    DocComment(String),
+
+    And,
+    Else,
+    If,
+    In,
+    Let,
+    Match,
+    Then,
+    Type,
+    With,
+
+    Colon,
+    Comma,
+    Dot,
+    Equals,
+    Lambda,
+    Pipe,
+    RArrow,
+
+    LBrace,
+    LBracket,
+    LParen,
+    RBrace,
+    RBracket,
+    RParen,
+
+    OpenBlock,
+    CloseBlock,
+    Semi,
+
+    EOF, // Required for the layout algorithm
+}
+
+impl<'input> fmt::Display for Token<'input> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Token::*;
+
+        let s = match *self {
+            Identifier(_) => "Identifier",
+            Operator(_) => "Operator",
+            StringLiteral(_) => "StringLiteral",
+            CharLiteral(_) => "CharLiteral",
+            IntLiteral(_) => "IntLiteral",
+            ByteLiteral(_) => "ByteLiteral",
+            FloatLiteral(_) => "FloatLiteral",
+            DocComment(_) => "DocComment",
+
+            And => "And",
+            Else => "Else",
+            If => "If",
+            In => "In",
+            Let => "Let",
+            Match => "Match",
+            Then => "Then",
+            Type => "Type",
+            With => "With",
+
+            LBrace => "LBrace",
+            LBracket => "LBracket",
+            LParen => "LParen",
+
+            RBrace => "RBrace",
+            RBracket => "RBracket",
+            RParen => "RParen",
+
+            Colon => "Colon",
+            Comma => "Comma",
+            Dot => "Dot",
+            Equals => "Equal",
+            Lambda => "Lambda",
+            Pipe => "Pipe",
+            RArrow => "RArrow",
+
+            OpenBlock => "OpenBlock",
+            CloseBlock => "CloseBlock",
+            Semi => "Semi",
+
+            EOF => "EOF",
+        };
+        s.fmt(f)
+    }
+}
+
+pub type SpannedToken<'input> = Spanned<Token<'input>, Location>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Error {
+    pub location: Location,
+    pub code: ErrorCode,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ErrorCode {
```

Comment:
I think we could use `quick_error` for this type? I don't think it is used in the result of parse_expr yet either?